dyne/examples.py

```diff
@@ -533,3 +534,155 @@ def h(k, X, with_jacobian=True):
+    P0 : array_like, shape (4, 4)
+        Initial covariance matrix.
+    q : array_like, shape (4,)
+        Additive process noise defined in a continuous sense.
+        Standard deviation of the discrete noise is computed as ``q * sqrt(dt)``, where
+        ``dt`` is `time_step`.
+    sigma_measurement : array_like, shape(3,)
+        Standard deviation of measurement noise.
+    rng : None, int or `numpy.random.RandomState`
+        Seed to create or already created RandomState. None (default) corresponds to
+        nondeterministic seeding.
+
+    Returns
+    -------
+    NonlinearProblemExample
+    """
+    deg2rad = np.deg2rad(1)
+
+    rng = check_random_state(rng)
+    n_epochs = np.round(total_time / time_step).astype(int)
+    time = time_step * np.arange(n_epochs)
+    phase = (2 * np.pi * time[:, None] / rph_change_period +
+             np.deg2rad(rph_change_phase_offset))
+    rph = (np.atleast_2d(rph_mean) +
+           np.atleast_2d(rph_change_amplitude) * np.sin(phase))
+    mn, me, md = magnetic_field
+
+    noises = np.asarray(q)
+    n_noises = np.sum(noises > 0)
+    G = np.zeros((4, n_noises))
+    Q = np.zeros((n_noises, n_noises))
+    j = 0
+    for i, s in enumerate(noises):
+        if s > 0:
+            G[i, j] = 1
+            Q[j, j] = s**2 * time_step
+            j = j + 1
+
+    def f(k, X, W=None, with_jacobian=True):
+        X = np.array(X)
+        X[0] += (rph[k+1, 2] - rph[k, 2])
+        F = np.eye(4, 4)
+        wk = G @ W if W is not None else 0
+        return (X + wk, F, G) if with_jacobian else X + wk
+
+    def h(k, X, with_jacobian=True):
+        rph_deg = np.array(rph[k])
+        rph_deg[2] = X[0]
+        C_nb = Rotation.from_euler('xyz', rph_deg, degrees=True).as_matrix()
+        Z = C_nb.T @ magnetic_field + X[1:]
+        rph_rad = deg2rad * rph_deg
+
+        sr = np.sin(rph_rad[0])
+        cr = np.cos(rph_rad[0])
+        sp = np.sin(rph_rad[1])
+        cp = np.cos(rph_rad[1])
+        sh = np.sin(rph_rad[2])
+        ch = np.cos(rph_rad[2])
+
+        H = np.zeros((3, 4))
+        H[0, 0] = me*ch*cp - mn*sh*cp
+        H[1, 0] = me*(-sh*cr + sp*sr*cr) + mn*(-sh*sp*sr - ch*cr)
+        H[2, 0] = me*( sh*sr + sp*ch*cr) + mn*(-sh*sp*cr + sr*ch)
+        H *= deg2rad
+        H[:, 1:] = np.eye(3)
+
+        return (Z, H) if with_jacobian else Z
+
+    X0t = np.zeros(4)
+    X0t[0] = rph[0, 2]
+    X0t[1:] = mag_bias
+
+    X = X0t
+    Xt = np.empty((n_epochs, 4))
+    Wt = np.empty((n_epochs - 1, n_noises))
+    R = np.diag(sigma_measurement)**2
+    Z = []
+    measurement_epochs = np.arange(n_epochs)
+
+    for k in range(n_epochs):
+        Xt[k] = X
+        vk = rng.multivariate_normal(np.zeros(len(R)), R)
+        Z.append(h(k, X, with_jacobian=False) + vk)
+
+        if k + 1 < n_epochs:
+            if len(Q) > 0:
+                Wt[k] = rng.multivariate_normal(np.zeros(len(Q)), Q)
+            X, *_ = f(k, X, Wt[k])
+
+    Q = np.array((n_epochs - 1) * [Q])
```

Comment:
Not necessary, I've already said that and it is documented for every estimation algorithm.